core/security_group_rule.go

```diff
@@ -0,0 +1,191 @@
+package core
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+
+	"github.com/krystal/go-katapult"
+)
+
+type SecurityGroupRule struct {
+	ID        string   `json:"id,omitempty"`
+	Direction string   `json:"direction,omitempty"`
+	Protocol  string   `json:"protocol,omitempty"`
+	Ports     string   `json:"ports,omitempty"`
+	Targets   []string `json:"targets,omitempty"`
+	Notes     string   `json:"notes,omitempty"`
+}
+
+func (sgr *SecurityGroupRule) Ref() SecurityGroupRuleRef {
+	return SecurityGroupRuleRef{ID: sgr.ID}
+}
+
+type SecurityGroupRuleRef struct {
+	ID string `json:"id,omitempty"`
+}
+
+func (sgr SecurityGroupRuleRef) queryValues() *url.Values {
+	v := &url.Values{}
+	v.Set("security_group_rule[id]", sgr.ID)
+
+	return v
+}
+
+type SecurityGroupRuleArguments struct {
+	Direction string   `json:"direction,omitempty"`
+	Protocol  string   `json:"protocol,omitempty"`
+	Ports     string   `json:"ports,omitempty"`
+	Targets   []string `json:"targets,omitempty"`
+	Notes     string   `json:"notes,omitempty"`
+}
+
+type SecurityGroupRulesResponseBody struct {
+	Pagination         *katapult.Pagination `json:"pagination,omitempty"`
+	SecurityGroupRule  *SecurityGroupRule   `json:"security_group_rule,omitempty"`
+	SecurityGroupRules []SecurityGroupRule  `json:"security_group_rules,omitempty"`
+}
+
+type SecurityGroupRulesClient struct {
+	client   RequestMaker
+	basePath *url.URL
+}
+
+// NewSecurityGroupRulesClient returns a new SecurityGroupRulesClient for
+// interacting with SecurityGroup Rules.
+func NewSecurityGroupRulesClient(rm RequestMaker) *SecurityGroupRulesClient {
+	return &SecurityGroupRulesClient{
+		client:   rm,
+		basePath: &url.URL{Path: "/core/v1/"},
+	}
+}
+
+// List returns SecurityGroup Rules for the specified SecurityGroup.
+func (s *SecurityGroupRulesClient) List(
+	ctx context.Context,
+	sg SecurityGroupRef,
+	opts *ListOptions,
+) ([]SecurityGroupRule, *katapult.Response, error) {
+	qs := queryValues(opts, sg)
+	u := &url.URL{
+		Path:     "security_groups/_/rules",
+		RawQuery: qs.Encode(),
+	}
+
+	body, resp, err := s.doRequest(ctx, "GET", u, nil)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	resp.Pagination = body.Pagination
+
+	return body.SecurityGroupRules, resp, err
+}
+
+func (s *SecurityGroupRulesClient) Get(
+	ctx context.Context,
+	ref SecurityGroupRuleRef,
+) (*SecurityGroupRule, *katapult.Response, error) {
+	u := &url.URL{
+		Path:     "security_groups/rules/_",
+		RawQuery: ref.queryValues().Encode(),
+	}
+	body, resp, err := s.doRequest(ctx, "GET", u, nil)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return body.SecurityGroupRule, resp, err
+}
+
+func (s *SecurityGroupRulesClient) GetByID(
+	ctx context.Context,
+	id string,
+) (*SecurityGroupRule, *katapult.Response, error) {
+	return s.Get(ctx, SecurityGroupRuleRef{ID: id})
+}
+
+type SecurityGroupRuleCreateRequest struct {
```

Comment:
This struct doesn't need to be exported, as it's only used for internal request purposes.